### PR TITLE
Support update documentation CA for included record params

### LINKED ISSUE #42304: [Bug]: `Update documentation` CA on a different parameter removes the documentation made for the `included-record-params`
### Description

$title.

https://github.com/ballerina-platform/ballerina-lang/assets/59343084/23772145-e73d-4ffc-b07b-473ee5b77c4d

### Steps to Reproduce

```ballerina 
type ParamConfig record {|
    string timeEntry;
    string msg;
|};

# Description.
#
# + id - parameter description  
# + config - parameter description
function logMsg(int id, int value, *ParamConfig config) {

}
```

### Affected Version(s)

Ballerina 2201.8.5 (Swan Lake Update 8)

### OS, DB, other environment details and versions

OS: macOS 14.2.1 23C71
JDK: openjdk 17.0.8 2023-07-18

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose

Fixes #42304

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/command/docs/DocumentationGenerator.java [hunk 1]
```diff
@@ -336,12 +337,15 @@ private static DocAttachmentInfo getFunctionNodeDocumentation(FunctionSignatureN
...
                 paramName = ((RequiredParameterNode) param).paramName();
-            } else if (param.kind() == SyntaxKind.DEFAULTABLE_PARAM) {
+            } else if (paramKind == SyntaxKind.DEFAULTABLE_PARAM) {
                 paramName = ((DefaultableParameterNode) param).paramName();
-            } else if (param.kind() == SyntaxKind.REST_PARAM) {
+            } else if (paramKind == SyntaxKind.REST_PARAM) {
                 paramName = ((RestParameterNode) param).paramName();
+            } else if (paramKind == SyntaxKind.INCLUDED_RECORD_PARAM) {
+                paramName = ((IncludedRecordParameterNode) param).paramName();
             }
```
Comment (nipunayf) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/command/docs/DocumentationGenerator.java:349]: Can replace all with one statement:

```java
Optional<Token> paramName = switch (param.kind()) {
    case REQUIRED_PARAM -> ((RequiredParameterNode) param).paramName();
    case DEFAULTABLE_PARAM -> ((DefaultableParameterNode) param).paramName();
    case REST_PARAM -> ((RestParameterNode) param).paramName();
    case INCLUDED_RECORD_PARAM -> ((IncludedRecordParameterNode) param).paramName();
    default -> Optional.empty();
};
```

**Review by nipunayf — COMMENTED** [review]
On a separate note, we can move `annotation`, `typeName`, and `paramName` to the abstract class `ParameterNode` as all its inherited classes implement them. With this approach, we can reduce the duplicate code.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/command/docs/DocumentationGenerator.java (hunk 1 above)
Comment (KavinduZoysa) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/command/docs/DocumentationGenerator.java]: Changed

**Review by nipunayf — COMMENTED** [review]
The PR fixes this issue https://github.com/ballerina-platform/ballerina-lang/issues/42303 as well. We can mention this in the PR after adding a couple of test cases:
1. Update documentation CA for a doc without the included param.
```ballerina
# Description.
#
# + id - parameter description  
# + value - parameter description
# + return - return value description
function logMsg(int id, int value, *ParamConfig config) returns int {
    return 0;
}
```

2. Add documentation CA for a function
```ballerina
function logMsg(int id, int value, *ParamConfig config) returns int {
    return 0;
}
```

**Review by LakshanWeerasinghe — APPROVED** [review]

**Comment by mindula** [comment]
For the scenario below, we are not getting the `Update document` code action for `*ParamConfig config`. Even the `Document this` code action does not add a description for `*ParamConfig config`. Shall we check the cause of that as well? Or maybe we can fix it with another issue as we will also have to change the `Document this` code action.

```
# Description.
#
# + id - parameter description  
# + value - parameter description
# + return - return value description
function logMsg(int id, int value, *ParamConfig config) returns int {
    return 0;
}
```

**Comment by nipunayf** [comment]
> For the scenario below, we are not getting the `Update document` code action for `*ParamConfig config`. Even the `Document this` code action does not add a description for `*ParamConfig config`. Shall we check the cause of that as well? Or maybe we can fix it with another issue as we will also have to change the `Document this` code action.
> 
> ```
> # Description.
> #
> # + id - parameter description  
> # + value - parameter description
> # + return - return value description
> function logMsg(int id, int value, *ParamConfig config) returns int {
>     return 0;
> }
> ```

This is tracked with https://github.com/ballerina-platform/ballerina-lang/issues/42303, and fixed with this PR:

https://github.com/ballerina-platform/ballerina-lang/assets/59343084/47d202dd-2414-42fd-8b73-d68f800dc88f

As mentioned in my previous comment, we need to add the respective test cases.

**Comment by mindula** [comment]
> > For the scenario below, we are not getting the `Update document` code action for `*ParamConfig config`. Even the `Document this` code action does not add a description for `*ParamConfig config`. Shall we check the cause of that as well? Or maybe we can fix it with another issue as we will also have to change the `Document this` code action.
> > ```
> > # Description.
> > #
> > # + id - parameter description  
> > # + value - parameter description
> > # + return - return value description
> > function logMsg(int id, int value, *ParamConfig config) returns int {
> >     return 0;
> > }
> > ```
> 
> This is tracked with #42303, and fixed with this PR:
> 
>  Screen.Recording.2024-05-28.at.11.07.54.AM.mov 
> As mentioned in my previous comment, we need to add the respective test cases.

I'm still not getting the code action and the document this code action does not add a field too, that's why I was concerned.

**Comment by mindula** [comment]
Another scenario I noticed was when we have the documentation for the above mentioned scenario, it still suggests the Update documentation code action and removes the documented field of `config`,

```
# Description.
#
# + id - parameter description  
# + value - parameter description
# + config - parameter description
# + return - return value description
function logMsg(int id, int value, *ParamConfig config) returns int {
    return 0;
}
```

**Comment by KavinduZoysa** [comment]
Will address comments in a different PR.

**Pull request merged**